grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,314 @@
+            return SERVING_RESPONSE;
+        } else if (status == ServingStatus.NOT_SERVING) {
+            return NOT_SERVING_RESPONSE;
+        } else if (status == ServingStatus.SERVICE_UNKNOWN) {
+            return SERVICE_UNKNOWN_RESPONSE;
+        } else {
+            throw new IllegalArgumentException("Invalid status:" + status);
+        }
+    }
+
+    private static void setHealthUpdateListenerToServerHealthCheckers(
+            Consumer<String> healthCheckUpdateListener,
+            Collection<ListenableHealthChecker> listenableHealthCheckers) {
+        listenableHealthCheckers.forEach(
+                lhc -> lhc.addListener(healthChecker -> healthCheckUpdateListener.accept("")));
+    }
+
+    private static void setHealthUpdateListenerToGrpcServiceHealthCheckers(
+            Consumer<String> healthCheckUpdateListener,
+            Iterable<Entry<String, ListenableHealthChecker>> listenableHealthCheckers) {
+        listenableHealthCheckers.forEach(entry -> entry.getValue().addListener(
+                healthChecker -> healthCheckUpdateListener.accept(entry.getKey())));
+    }
+
+    private final SettableHealthChecker serverHealth;
+    private final Set<ListenableHealthChecker> serverHealthCheckers;
+    private final Map<String, ListenableHealthChecker> grpcServiceHealthCheckers;
+    private final Multimap<String, Entry<StreamObserver<HealthCheckResponse>, ServingStatus>> watchers =
+            HashMultimap.create();
+    @Nullable
+    private Server server;
+
+    GrpcHealthCheckService(
+            Set<ListenableHealthChecker> serverHealthCheckers,
+            Map<String, ListenableHealthChecker> grpcServiceHealthCheckers,
+            List<HealthCheckUpdateListener> updateListeners
+    ) {
+        serverHealth = new SettableHealthChecker(false);
+        if (!updateListeners.isEmpty()) {
+            addServerHealthUpdateListener(updateListeners);
+        }
+        this.serverHealthCheckers = ImmutableSet.<ListenableHealthChecker>builder()
+                                                .add(serverHealth)
+                                                .addAll(serverHealthCheckers)
+                                                .build();
+        this.grpcServiceHealthCheckers = grpcServiceHealthCheckers;
+        final Consumer<String> healthCheckUpdateListener = provideInternalHealthUpdateListener();
+        setHealthUpdateListenerToServerHealthCheckers(healthCheckUpdateListener, this.serverHealthCheckers);
+        setHealthUpdateListenerToGrpcServiceHealthCheckers(healthCheckUpdateListener,
+                                                           this.grpcServiceHealthCheckers.entrySet());
+    }
+
+    @Override
+    public void check(HealthCheckRequest request, StreamObserver<HealthCheckResponse> responseObserver) {
+        final String service = request.getService();
+        final ServingStatus status = checkServingStatus(service);
+        if (status == ServingStatus.SERVICE_UNKNOWN) {
```

Comment:
The doc says
`The server should register all the services manually and set the individual status, ...`.
But the current implementation returns `SERVICE_UNKNOWN` if the corresponding `HealthChecker` is not registered via `checkerForGrpcService(...)`.

In order to solve this, I think we need to add all service names to `GrpcHealthCheckService` in `GrpcServiceBuilder.build()`.
Then, when a check request is sent from the client, we can check if the service name is valid or not and use the server health if the corresponding `HealthChecker` of the service is not registered.

Reply:
I thought `manually` means that users need to set the individual status.
I've checked gRPC-Java's implementation and couldn't find a logic that automatically registers all service names.
The following API is open to users to set the status of each service.
https://github.com/grpc/grpc-java/blob/bab1fe38dc50d4178955b79cfb1636afd2aa64df/services/src/main/java/io/grpc/protobuf/services/HealthStatusManager.java#L65

Reply:
Thanks! Let's leave it as is. 😄